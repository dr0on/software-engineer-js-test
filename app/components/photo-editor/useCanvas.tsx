import { useState, useEffect, useRef } from 'react';

export const enum ImageAction {
  Top,
  Down,
  Left,
  Right,
  ZoomIn,
  ZoomOut,
}

const useCanvas = () => {
  const [image, setImage] = useState<HTMLImageElement>(new Image());
  const [options, setOptions] = useState({ x: 0, y: 0, scale: 1 });
  const canvasRef = useRef<HTMLCanvasElement>(null);

  useEffect(() => {
    if (image && canvasRef.current) {
      const canvas = canvasRef.current;
      const ctx = canvas.getContext('2d');

      ctx?.clearRect(0, 0, canvas.width, canvas.height);
      ctx?.drawImage(
        image,
        options.x,
        options.y,
        image.naturalWidth * options.scale,
        image.naturalHeight * options.scale
      );
    }
  }, [image, options.x, options.y, options.scale]);

  const calculateZoomOptions = (img: any, scale: number) => {
    const currentWidth = img.width * options.scale;
    const currentHeight = img.height * options.scale;

    const newWidth = currentWidth * scale;
    const newHeight = currentHeight * scale;

    const xOffset = (newWidth - currentWidth) / 2;
    const yOffset = (newHeight - currentHeight) / 2;

    return {
      x: options.x - xOffset,
      y: options.y - yOffset,
      scale: options.scale * scale,
    };
  };

  const isBottomLimit = () => {
    const canvasHeight = canvasRef.current?.height || 0;
    const scaledHeight = image.height * options.scale;

    return options.y + scaledHeight - 10 < canvasHeight;
  };

  const isLeftLimit = () => {
    const canvasWidth = canvasRef.current?.width || 0;
    const scaledWidth = image.width * options.scale;

    return options.x + scaledWidth - 10 < canvasWidth;
  };

  const isTopLimit = () => options.y + 10 > 0;

  const isRightLimit = () => options.x + 10 > 0;

  const setZoomIn = () => {
    const { x, y, scale } = calculateZoomOptions(image, 1.1);

    setOptions({ x, y, scale });
  };

  const setZoomOut = () => {
    if (!isTopLimit() && !isBottomLimit() && !isLeftLimit() && !isRightLimit()) {
      const { x, y, scale } = calculateZoomOptions(image, 0.9);

      setOptions({ x, y, scale });
    }
  };

  const setMoveTop = () => {
    const y = isTopLimit() ? 0 : options.y + 10;
    setOptions({ x: options.x, y, scale: options.scale });
  };

  const setMoveBottom = () => {
    const canvasHeight = canvasRef.current?.height || 0;
    let y = isBottomLimit() ? canvasHeight - image.height * options.scale : options.y - 10;

    setOptions({ x: options.x, y, scale: options.scale });
  };

  const setMoveRight = () => {
    const x = isRightLimit() ? 0 : options.x + 10;

    setOptions({ x, y: options.y, scale: options.scale });
  };

  const setMoveLeft = () => {
    const canvasWidth = canvasRef.current?.width || 0;
    let x = isLeftLimit() ? canvasWidth - image.width * options.scale : options.x - 10;

    setOptions({ x, y: options.y, scale: options.scale });
  };

  const setImageAction = (action: ImageAction) => {
    switch (action) {
      case ImageAction.ZoomIn:
        setZoomIn();
        break;
      case ImageAction.ZoomOut:
        setZoomOut();
        break;
      case ImageAction.Top:
        setMoveTop();
        break;
      case ImageAction.Down:
        setMoveBottom();
        break;
      case ImageAction.Right:
        setMoveRight();
        break;
      case ImageAction.Left:
        setMoveLeft();
        break;
    }
  };

  return { canvasRef, image, setImage, options, setOptions, setImageAction };
};

export default useCanvas;
