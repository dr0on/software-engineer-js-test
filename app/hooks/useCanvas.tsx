import { useState, useEffect, useRef } from 'react';

export const enum ImageAction {
  Top,
  Down,
  Left,
  Right,
  ZoomIn,
  ZoomOut,
}

const MOVE_STEP = 10;

const useCanvas = () => {
  const [image, setImage] = useState<HTMLImageElement>(new Image());
  const [options, setOptions] = useState({ x: 0, y: 0, scale: 1 });
  const canvasRef = useRef<HTMLCanvasElement>(null);

  const canvasHeight = canvasRef.current?.height || 0;
  const canvasWidth = canvasRef.current?.width || 0;

  useEffect(() => {
    if (image && canvasRef.current) {
      const canvas = canvasRef.current;
      const ctx = canvas.getContext('2d');

      ctx?.clearRect(0, 0, canvas.width, canvas.height);
      ctx?.drawImage(
        image,
        options.x,
        options.y,
        image.naturalWidth * options.scale,
        image.naturalHeight * options.scale
      );
    }
  }, [image, options.x, options.y, options.scale]);

  const calculateZoomOptions = (img: any, scale: number) => {
    const currentWidth = img.width * options.scale;
    const currentHeight = img.height * options.scale;

    const newWidth = currentWidth * scale;
    const newHeight = currentHeight * scale;

    const xOffset = (newWidth - currentWidth) / 2;
    const yOffset = (newHeight - currentHeight) / 2;

    return {
      x: options.x - xOffset,
      y: options.y - yOffset,
      scale: options.scale * scale,
    };
  };

  const isBottomLimit = () => {
    const scaledHeight = image.height * options.scale;
    return options.y + scaledHeight - MOVE_STEP < canvasHeight;
  };

  const isLeftLimit = () => {
    const scaledWidth = image.width * options.scale;
    return options.x + scaledWidth - MOVE_STEP < canvasWidth;
  };

  const isTopLimit = () => options.y + MOVE_STEP > 0;

  const isRightLimit = () => options.x + MOVE_STEP > 0;

  const setZoomIn = () => {
    const { x, y, scale } = calculateZoomOptions(image, 1.1);
    setOptions({ x, y, scale });
  };

  const setZoomOut = () => {
    if (!isTopLimit() && !isBottomLimit() && !isLeftLimit() && !isRightLimit()) {
      const { x, y, scale } = calculateZoomOptions(image, 0.9);
      setOptions({ x, y, scale });
    }
  };

  const setMoveTop = () => {
    const y = isTopLimit() ? 0 : options.y + MOVE_STEP;
    setOptions({ x: options.x, y, scale: options.scale });
  };

  const setMoveBottom = () => {
    let y = isBottomLimit() ? canvasHeight - image.height * options.scale : options.y - MOVE_STEP;
    setOptions({ x: options.x, y, scale: options.scale });
  };

  const setMoveRight = () => {
    const x = isRightLimit() ? 0 : options.x + MOVE_STEP;
    setOptions({ x, y: options.y, scale: options.scale });
  };

  const setMoveLeft = () => {
    let x = isLeftLimit() ? canvasWidth - image.width * options.scale : options.x - MOVE_STEP;
    setOptions({ x, y: options.y, scale: options.scale });
  };

  const setImageAction = (action: ImageAction) => {
    switch (action) {
      case ImageAction.ZoomIn:
        setZoomIn();
        break;
      case ImageAction.ZoomOut:
        setZoomOut();
        break;
      case ImageAction.Top:
        setMoveTop();
        break;
      case ImageAction.Down:
        setMoveBottom();
        break;
      case ImageAction.Right:
        setMoveRight();
        break;
      case ImageAction.Left:
        setMoveLeft();
        break;
    }
  };

  return { canvasRef, image, setImage, options, setOptions, setImageAction };
};

export default useCanvas;
