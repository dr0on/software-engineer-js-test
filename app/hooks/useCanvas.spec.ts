/**
 * @jest-environment jsdom
 */

import React  from 'react';
import 'jest-canvas-mock';
import { renderHook, act } from '@testing-library/react-hooks';

import useCanvas, { ImageAction } from './useCanvas';

let canvas: HTMLCanvasElement;
let ctx: CanvasRenderingContext2D | null;

describe('useCanvas', () => {
  beforeEach(() => {
    canvas = document.createElement('canvas');
    ctx = canvas.getContext('2d');
    canvas.width = 400;
    canvas.height = 300;

    jest.spyOn(React, 'useRef').mockReturnValue({ current: canvas });
  });

  it('should return default options', () => {
    jest.spyOn(React, 'useState')
      .mockImplementationOnce(() => [new Image(), () => null])
      .mockImplementationOnce(() => [{ x: 0, y: 0, scale: 1 }, () => null]);
    jest.spyOn(React, 'useRef').mockReturnValueOnce({ current: canvas });
    jest.spyOn(React, 'useEffect').mockImplementation(f => f());

    const { options } = useCanvas();

    expect(options).toEqual({ x: 0, y: 0, scale: 1 });
  });

  it('should return canvas ref', () => {
    jest.spyOn(React, 'useState')
      .mockImplementationOnce(() => [new Image(), () => null])
      .mockImplementationOnce(() => [{ x: 0, y: 0, scale: 1 }, () => null]);
    jest.spyOn(React, 'useRef').mockReturnValueOnce({ current: canvas });
    jest.spyOn(React, 'useEffect').mockImplementation(f => f());

    const { canvasRef } = useCanvas();

    expect(canvasRef.current).toEqual(canvas);
  });

  it('should return image', () => {
    jest.spyOn(React, 'useState')
      .mockImplementationOnce(() => [new Image(200, 100), () => null])
      .mockImplementationOnce(() => [{ x: 0, y: 0, scale: 1 }, () => null]);
    jest.spyOn(React, 'useRef').mockReturnValueOnce({ current: canvas });
    jest.spyOn(React, 'useEffect').mockImplementation(f => f());

    const { image } = useCanvas();

    expect(image.width).toEqual(200);
  });

  it('should zoom out image', () => {
    const { result } = renderHook(() => useCanvas());

    act(() => { result.current.setOptions({ x: -50, y: -50, scale: 1 }) });
    act(() => { result.current.setImage(new Image(700, 700)) });
    act(() => { result.current.setImageAction(ImageAction.ZoomOut) });

    const events = ctx?.__getEvents();

    expect(events).toMatchSnapshot();
    expect(result.current.options).toMatchObject({ x: -15, y: -15, scale: 0.9 });
  });

  it('should zoom in image', () => {
    const { result } = renderHook(() => useCanvas());

    act(() => { result.current.setOptions({ x: -50, y: -50, scale: 1 }) });
    act(() => { result.current.setImage(new Image(700, 700)) });
    act(() => { result.current.setImageAction(ImageAction.ZoomIn) });

    const events = ctx?.__getEvents();

    expect(events).toMatchSnapshot();
    expect(result.current.options).toMatchObject({ x: -85.00000000000006, y: -85.00000000000006, scale: 1.1 });
  });

  it('should move image to top', () => {
    const { result } = renderHook(() => useCanvas());

    act(() => { result.current.setOptions({ x: 0, y: -15, scale: 1 }) });
    act(() => { result.current.setImage(new Image(700, 700)) });
    act(() => { result.current.setImageAction(ImageAction.Top) });

    const events = ctx?.__getEvents();

    expect(events).toMatchSnapshot();
    expect(result.current.options).toMatchObject({ x: 0, y: -5, scale: 1 });
  });

  it('should move image bottom', () => {
    const { result } = renderHook(() => useCanvas());

    act(() => { result.current.setOptions({ x: 0, y: -15, scale: 1 }) });
    act(() => { result.current.setImage(new Image(700, 700)) });
    act(() => { result.current.setImageAction(ImageAction.Down) });

    const events = ctx?.__getEvents();

    expect(events).toMatchSnapshot();
    expect(result.current.options).toMatchObject({ x: 0, y: -25, scale: 1 });
  });

  it('should move image left', () => {
    const { result } = renderHook(() => useCanvas());

    act(() => { result.current.setOptions({ x: 15, y: 0, scale: 1 }) });
    act(() => { result.current.setImage(new Image(700, 700)) });
    act(() => { result.current.setImageAction(ImageAction.Left) });

    const events = ctx?.__getEvents();

    expect(events).toMatchSnapshot();
    expect(result.current.options).toMatchObject({ x: 5, y: 0, scale: 1 });
  });

  it('should move image right', () => {
    const { result } = renderHook(() => useCanvas());

    act(() => { result.current.setOptions({ x: -15, y: 0, scale: 1 }) });
    act(() => { result.current.setImage(new Image(700, 700)) });
    act(() => { result.current.setImageAction(ImageAction.Right) });

    const events = ctx?.__getEvents();

    expect(events).toMatchSnapshot();
    expect(result.current.options).toMatchObject({ x: -5, y: 0, scale: 1 });
  });
});