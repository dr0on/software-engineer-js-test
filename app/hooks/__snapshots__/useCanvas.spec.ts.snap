// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useCanvas should move image bottom 1`] = `
[
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 0,
      "dy": 0,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 0,
      "dy": -15,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 700,
      "dWidth": 700,
      "dx": 0,
      "dy": -15,
      "img": <img
        height="700"
        width="700"
      />,
      "sHeight": 700,
      "sWidth": 700,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 700,
      "dWidth": 700,
      "dx": 0,
      "dy": -25,
      "img": <img
        height="700"
        width="700"
      />,
      "sHeight": 700,
      "sWidth": 700,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`useCanvas should move image left 1`] = `
[
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 0,
      "dy": 0,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 15,
      "dy": 0,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 700,
      "dWidth": 700,
      "dx": 15,
      "dy": 0,
      "img": <img
        height="700"
        width="700"
      />,
      "sHeight": 700,
      "sWidth": 700,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 700,
      "dWidth": 700,
      "dx": 5,
      "dy": 0,
      "img": <img
        height="700"
        width="700"
      />,
      "sHeight": 700,
      "sWidth": 700,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`useCanvas should move image right 1`] = `
[
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 0,
      "dy": 0,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -15,
      "dy": 0,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 700,
      "dWidth": 700,
      "dx": -15,
      "dy": 0,
      "img": <img
        height="700"
        width="700"
      />,
      "sHeight": 700,
      "sWidth": 700,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 700,
      "dWidth": 700,
      "dx": -5,
      "dy": 0,
      "img": <img
        height="700"
        width="700"
      />,
      "sHeight": 700,
      "sWidth": 700,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`useCanvas should move image to top 1`] = `
[
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 0,
      "dy": 0,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 0,
      "dy": -15,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 700,
      "dWidth": 700,
      "dx": 0,
      "dy": -15,
      "img": <img
        height="700"
        width="700"
      />,
      "sHeight": 700,
      "sWidth": 700,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 700,
      "dWidth": 700,
      "dx": 0,
      "dy": -5,
      "img": <img
        height="700"
        width="700"
      />,
      "sHeight": 700,
      "sWidth": 700,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`useCanvas should zoom in image 1`] = `
[
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 0,
      "dy": 0,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -50,
      "dy": -50,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 700,
      "dWidth": 700,
      "dx": -50,
      "dy": -50,
      "img": <img
        height="700"
        width="700"
      />,
      "sHeight": 700,
      "sWidth": 700,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 700,
      "dWidth": 700,
      "dx": -85.00000000000006,
      "dy": -85.00000000000006,
      "img": <img
        height="700"
        width="700"
      />,
      "sHeight": 700,
      "sWidth": 700,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`useCanvas should zoom out image 1`] = `
[
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 0,
      "dy": 0,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -50,
      "dy": -50,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 700,
      "dWidth": 700,
      "dx": -50,
      "dy": -50,
      "img": <img
        height="700"
        width="700"
      />,
      "sHeight": 700,
      "sWidth": 700,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "height": 300,
      "width": 400,
      "x": 0,
      "y": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  {
    "props": {
      "dHeight": 700,
      "dWidth": 700,
      "dx": -15,
      "dy": -15,
      "img": <img
        height="700"
        width="700"
      />,
      "sHeight": 700,
      "sWidth": 700,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;
